hub/core/chunk.py

```diff
@@ -110,29 +193,65 @@ def register_sample_to_headers(
             ValueError: If `incoming_num_bytes` is not divisible by `num_samples`.
         """
 
-        num_bytes_per_sample = incoming_num_bytes
         self.shapes_encoder.register_samples(sample_shape, 1)
-        self.byte_positions_encoder.register_samples(num_bytes_per_sample, 1)
+        if (
+            incoming_num_bytes is not None
+        ):  # incoming_num_bytes is not applicable for non lz4 compressions
+            self.byte_positions_encoder.register_samples(incoming_num_bytes, 1)
+        self._clear_decompressed_cache()
 
     def update_sample(
-        self, local_sample_index: int, new_buffer: memoryview, new_shape: Tuple[int]
+        self,
+        local_sample_index: int,
+        new_buffer: memoryview,
+        new_shape: Tuple[int],
+        chunk_compression: Optional[str] = None,
+        dtype: Optional[np.dtype] = np.dtype("uint8"),
     ):
         """Updates data and headers for `local_sample_index` with the incoming `new_buffer` and `new_shape`."""
 
         expected_dimensionality = len(self.shapes_encoder[local_sample_index])
         if expected_dimensionality != len(new_shape):
             raise TensorInvalidSampleShapeError(new_shape, expected_dimensionality)
-
         new_nb = len(new_buffer)
+        self.shapes_encoder[local_sample_index] = new_shape
+        if chunk_compression:
+            if chunk_compression == "lz4":
+                decompressed_buffer = self.decompressed_data()
+                old_start_byte, old_end_byte = self.byte_positions_encoder[
+                    local_sample_index
+                ]
+
+                left = decompressed_buffer[:old_start_byte]
+                right = decompressed_buffer[old_end_byte:]
+                total_new_bytes = len(left) + new_nb + len(right)
+                new_data_uncompressed = bytearray(total_new_bytes)
+                self.byte_positions_encoder[local_sample_index] = new_nb
+                new_start_byte, new_end_byte = self.byte_positions_encoder[
+                    local_sample_index
+                ]
+                new_data_uncompressed[:new_start_byte] = left
+                new_data_uncompressed[new_start_byte:new_end_byte] = new_buffer
+                new_data_uncompressed[new_end_byte:] = right
+                self._data = memoryview(lz4.frame.compress(new_data_uncompressed))
+                self._decompressed_data = memoryview(new_data_uncompressed)
+            else:
+                decompressed_samples = self.decompressed_samples()
+                decompressed_samples[local_sample_index] = np.frombuffer(
```

Comment:
it would be very helpful to break these into functions and/or add comments. it's pretty hard to follow (moving it into a modular structure would help readability)

Reply:
Added comments.